pkg/quantile/bin.go

```diff
@@ -0,0 +1,86 @@
+package quantile
+
+import (
+	"math"
+	"strings"
+)
+
+const (
+	maxBinWidth = math.MaxUint16
+)
+
+type bin struct {
+	k Key
+	n uint16
+}
+
+// incrSafe performs `b.n += by` safely handling overflows. When an overflow
+// occurs, we set b.n to it's max, and return the leftover amount to increment.
+func (b *bin) incrSafe(by int) int {
+	next := by + int(b.n)
+
+	if next > maxBinWidth {
+		b.n = maxBinWidth
+		return next - maxBinWidth
+	}
+
+	b.n = uint16(next)
+	return 0
+}
+
+// appendSafe appends 1 or more bins with the given key safely handing overflow by
+// inserting multiple buckets when needed.
+//   (1) n <= maxBinWidth :  1 bin
+//   (2) n > maxBinWidth  : >1 bin
+func appendSafe(bins []bin, k Key, n int) []bin {
+	if n <= maxBinWidth {
+		return append(bins, bin{k: k, n: uint16(n)})
+	}
+
+	// on overflow, insert multiple bins with the same key.
+	// put full bins at end
+
+	// TODO|PROD: Add validation func that sorts by key and then n (smaller bin first).
+	r := uint16(n % maxBinWidth)
+	if r != 0 {
+		bins = append(bins, bin{k: k, n: r})
+	}
+
+	for i := 0; i < int(n/maxBinWidth); i++ {
+		bins = append(bins, bin{k: k, n: maxBinWidth})
+	}
+
+	return bins
+}
+
+type binList []bin
+
+func (bins binList) nSum() int {
+	s := 0
+	for _, b := range bins {
+		s += int(b.n)
+	}
+	return s
+}
+
+func (bins binList) Cap() int {
+	return cap(bins)
+}
+
+func (bins binList) Len() int {
+	return len(bins)
+}
+
+func (bins binList) ensureLen(newLen int) binList {
```

Comment:
Is there a reason for the iterative appends vs reallocating and copying like:
```
if cap(bins) < newLen{
  newBins = make(binList, newLen)
  copy(newBins, bins)
  return newbins
}
return bins[:newLen]
```
This also has the convenient property of providing zero values for the elements `len(bins)` through `cap(bins)-1`, although I'm not sure if that matters.

Reply:
This will eventually be replaced with a sized tiered `sync.Pool` but mainly because this idiom is faster in go1.10.

https://github.com/golang/go/issues/21266

Reply:
`ensureLen` is called from:

1. `store.merge`
https://github.com/DataDog/datadog-agent/blob/45fe8409a7569dac247591ae3ea34bc27f7a7d01/pkg/quantile/store.go#L107-L137
- `s.bins` grows incrementally here, so it is much faster to append a small number of elements and waste a bit of memory (luckily since a bin is only 32 bytes we don't waste much).
- size tiered sync.Pool would work well here.

2. `sketch.Copy`
https://github.com/DataDog/datadog-agent/blob/45fe8409a7569dac247591ae3ea34bc27f7a7d01/pkg/quantile/sparse.go#L126-L133
- re-slicing would be better here